src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs

```diff
@@ -16,58 +17,47 @@ namespace Generators
 {
     public partial class EventSourceGenerator
     {
-        private sealed class Parser
+        private static class Parser
         {
-            private readonly CancellationToken _cancellationToken;
-            private readonly Compilation _compilation;
-            private readonly Action<Diagnostic> _reportDiagnostic;
-
-            public Parser(Compilation compilation, Action<Diagnostic> reportDiagnostic, CancellationToken cancellationToken)
-            {
-                _compilation = compilation;
-                _cancellationToken = cancellationToken;
-                _reportDiagnostic = reportDiagnostic;
-            }
-
-            public EventSourceClass[] GetEventSourceClasses(List<ClassDeclarationSyntax> classDeclarations)
+            public static ImmutableArray<EventSourceClass> GetEventSourceClasses(ImmutableArray<ClassDeclarationSyntax> classDeclarations, Compilation compilation, CancellationToken cancellationToken)
             {
-                INamedTypeSymbol? autogenerateAttribute = _compilation.GetBestTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAutoGenerateAttribute");
+                INamedTypeSymbol? autogenerateAttribute = compilation.GetBestTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAutoGenerateAttribute");
                 if (autogenerateAttribute is null)
                 {
                     // No EventSourceAutoGenerateAttribute
-                    return Array.Empty<EventSourceClass>();
+                    return ImmutableArray<EventSourceClass>.Empty;
                 }
 
-                INamedTypeSymbol? eventSourceAttribute = _compilation.GetBestTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAttribute");
+                INamedTypeSymbol? eventSourceAttribute = compilation.GetBestTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAttribute");
                 if (eventSourceAttribute is null)
                 {
                     // No EventSourceAttribute
-                    return Array.Empty<EventSourceClass>();
+                    return ImmutableArray<EventSourceClass>.Empty;
                 }
 
-                List<EventSourceClass>? results = null;
+                ImmutableArray<EventSourceClass>.Builder? results = null;
                 // we enumerate by syntax tree, to minimize the need to instantiate semantic models (since they're expensive)
-                foreach (IGrouping<SyntaxTree, ClassDeclarationSyntax>? group in classDeclarations.GroupBy(x => x.SyntaxTree))
+                foreach (IGrouping<SyntaxTree, ClassDeclarationSyntax> group in classDeclarations.GroupBy(static x => x.SyntaxTree))
```

Comment:
Ideally, you would move this work into the semantic check part of the SyntaxProvider. That way you don't need to new up a semantic model (there already is one) which allows it to be shared among other generators.